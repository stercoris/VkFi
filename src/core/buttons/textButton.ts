import { ButtonColor, IKeyboardProxyButton, Keyboard } from "vk-io";

export interface IButton {
  label: string;
  color: ButtonColor;
  payload?: {
    command?: string;
    action?: string;
  };
}

type CreateTextButtonFunc<Props> = (args: Props) => IButton | Promise<IButton>;

export class TextButton<Props = void> {
  public create: Props extends void
    ? () => IKeyboardProxyButton | Promise<IKeyboardProxyButton>
    : (props: Props) => IKeyboardProxyButton | Promise<IKeyboardProxyButton>;

  private RawCreateButton: (props: Props) => IButton | Promise<IButton>;
  public Action: string;
  protected Command: string;

  constructor(createFun: CreateTextButtonFunc<Props> | IButton) {
    const postCreateButton =
      typeof createFun === "function" ? createFun : () => createFun;

    this.RawCreateButton = postCreateButton;

    //@ts-ignore
    this.create = async (props: Props) => {
      const button = await this.RawCreateButton(props);

      Object.assign(button, {
        payload: {
          action: this.Action,
          command: this.Command,
        },
      });

      return Keyboard.textButton(button);
    };
  }
}
