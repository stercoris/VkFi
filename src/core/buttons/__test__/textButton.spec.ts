import { ButtonColor } from "vk-io";
import { TextButton } from "../textButton";

describe("Text Button", () => {
  const Button = new TextButton(() => {
    return {
      label: "Sas",
      color: ButtonColor.NEGATIVE,
    };
  });

  const PropsButton = new TextButton((props: {label: string}) => ({
    label: props.label,
    color: ButtonColor.NEGATIVE
    
  }))

  it("Should work without parameters", async () => {
    expect(await Button.create()).toMatchSnapshot();
  })

  it("Should work with parameters", async () => {
    expect(await PropsButton.create({ label: "Super sus"})).toMatchSnapshot();
  })

  it("Should generate correct button on 'Action' change", async () => {
    Button.Action = "SAS ACTION";
    expect(await Button.create()).toMatchSnapshot();
  });

  it("Should generate correct button on 'Action' change", async () => {
    Button.Action = "SAS ACTION";
    expect(await Button.create()).toMatchSnapshot();
  });
});
