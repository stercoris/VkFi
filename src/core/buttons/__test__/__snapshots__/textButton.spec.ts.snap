// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Button Should generate correct button on 'Action' change 1`] = `
Object {
  "kind": "text",
  "options": Object {
    "color": "negative",
    "label": "Sas",
    "payload": Object {
      "action": "SAS ACTION",
      "command": undefined,
    },
  },
}
`;

exports[`Text Button Should generate correct button on 'Action' change 2`] = `
Object {
  "kind": "text",
  "options": Object {
    "color": "negative",
    "label": "Sas",
    "payload": Object {
      "action": "SAS ACTION",
      "command": undefined,
    },
  },
}
`;

exports[`Text Button Should work with parameters 1`] = `
Object {
  "kind": "text",
  "options": Object {
    "color": "negative",
    "label": "Super sus",
    "payload": Object {
      "action": undefined,
      "command": undefined,
    },
  },
}
`;

exports[`Text Button Should work without parameters 1`] = `
Object {
  "kind": "text",
  "options": Object {
    "color": "negative",
    "label": "Sas",
    "payload": Object {
      "action": undefined,
      "command": undefined,
    },
  },
}
`;
